src/librustc_typeck/check/op.rs

```diff
@@ -318,18 +330,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                     // This has nothing here because it means we did string
                                     // concatenation (e.g., "Hello " += "World!"). This means
                                     // we don't want the note in the else clause to be emitted
-                                } else if let ty::Param(p) = lhs_ty.kind {
-                                    suggest_constraining_param(
-                                        self.tcx,
-                                        self.body_id,
-                                        &mut err,
-                                        lhs_ty,
-                                        rhs_ty,
-                                        missing_trait,
-                                        p,
-                                        false,
-                                    );
-                                } else if !suggested_deref {
+                                    sugg = true;
+                                } else if let [ty] = &visitor.0[..] {
+                                    if let ty::Param(p) = ty.kind {
+                                        // FIXME: This *guesses* that constraining the type param
+                                        // will make the operation available, but this is only true
+                                        // when the corresponding trait has a blanked
+                                        // implementation, like the following:
+                                        // `impl<'a> PartialEq for &'a [T] where T: PartialEq {}`
+                                        // The correct thing to do would be to verify this
+                                        // projection would hold.
+                                        if *ty != lhs_ty {
+                                            note(&mut err, missing_trait);
+                                        }
+                                        suggest_constraining_param(
+                                            self.tcx,
+                                            self.body_id,
+                                            &mut err,
+                                            ty,
+                                            rhs_ty,
+                                            missing_trait,
+                                            p,
+                                            false,
+                                        );
+                                        sugg = true;
+                                    }
+                                }
+                                if !sugg && !suggested_deref {
```

Comment:
Is `sugg` necessary, couldn't this have remained `else if !suggested_deref`?

Reply:
I was thinking that if the single type found by the visitor wasn't a type param we wouldn't show the suggestion, but I'm seeing that the visitor is only looking at type params so the if let would be irrefutable. Changing.